### ISSUE TITLE
Warning maximum key length for a clustered index

### ISSUE THREAD

**Issue by thdotnet** [aspnet/Identity#1451]
I'm running SQL script to create the aspnet identity tables, but I'm getting the following warnings on Management Studio:

> 
> Warning! The maximum key length for a clustered index is 900 bytes. The index 'PK_AspNetUserLogins' has maximum length of 1800 bytes. For some combination of large values, the insert/update operation will fail.
> Warning! The maximum key length for a clustered index is 900 bytes. The index 'PK_AspNetUserRoles' has maximum length of 1800 bytes. For some combination of large values, the insert/update operation will fail.
> Warning! The maximum key length for a clustered index is 900 bytes. The index 'PK_AspNetUserTokens' has maximum length of 2700 bytes. For some combination of large values, the insert/update operation will fail.

Is that right? Is there any fix for that?

**Comment by blowdart** [comment]
What sql script? There's no sql script in ASP.NET Core identity. Do you perhaps mean the one in the .NET framework?

**Comment by thdotnet** [comment]
My bad, It's missing some important details. I want to use Identity without the Code First Approach. So what I did. I've created a sample app using the code first approach and let it creates the identity tables. Using SQL Management studio, I've generated the script from the previous step in order to create the tables on my real app (the one that does not use Code First).

It generates the errors described in the original description.

**Comment by Janidbest** [comment]
@thdotnet  do you mind elaborating on the missing important details? 
I have come across this issue today. Though it is a code first approach. As a part of our deployment plan, we are exporting the scripts from the migrations into a file and handing it over to DB admin so  this way we can prevent the application db access user will have low permissions.

**Comment by ajcvickers** [comment]
I will investigate this.

**Comment by Wesam18** [comment]
I am experiencing the same problem 
![image](https://user-images.githubusercontent.com/5215710/32058201-06c8f338-ba2f-11e7-99e7-d94bd4709179.png)

**Comment by thdotnet** [comment]
@Janidbest I already did that:

> So what I did: I've created a sample app using the code first approach and let it creates the identity tables. Using SQL Management studio, I've generated the script from the previous step in order to create the tables on my real app (the one that does not use Code First).
> 
> It generates the errors described in the original description.

@Wesam18  is facing the same problem

**Comment by Triwaters** [comment]
I noted this problem in my own work and got curious enough to search out this conversation.

I'm new to ASP.NET Identity, but not new to SQL Server. 
The three errors noted (all of which I see every time I propagate tables, etc. to a new database) are being thrown because ASP.NET Identity is breaking SQL Server rules. As SQL Server keeps patiently explaining every time one builds the ASP.Net Identity tables in their databases, a clustered index (usually that's your primary key on a table) cannot exceed 900 bytes.

That doesn't mean you can't create the index. What it means is that you may find yourself creating a legitimate and otherwise legal record that SQL Server nevertheless cannot save into the table. 

Meaning... 

The table design on AspNetUserLogins prescribes a primary key that is a composite of two fields: LoginProvider and ProviderKey. Both fields are nvarchar(450). That's two-bytes per character, and therefore one would think that it'd be possible to provide a full 450 unicode value in each field without breaking table rules for those fields....

...Except that when the record goes to save, the clustered index finds itself indexing 450x2x2 characters (two unicode fields of 450 chars) ... and craps out. That's 1,800 bytes. No can do. The index cannot hold that much data for a single record. You will not be able to save the record.

In practice? I'm betting this never happens. As I said, I am new to ASP.Net Identity, but I'd bet high odds these fields are usually quite short. So in practice, this is probably a non-problem.

But in principle, it's a pretty silly situation. Why has it not been remedied with more reasonable limits on field sizes?

I'm also wondering if the offending fields used to be varchar() and not nvarchar(). The problem goes away if they were varchar() fields--for two tables, at least. For AspNetUserTokens, even that wouldn't fix the problem. It's got a primary key that is a composite of three (3) nvarchar(450) fields. That's 2,700 bytes.

**Comment by Triwaters** [comment]
Not being one to leave "problems waiting to happen" (however remote) in my databases, not least because they usually surface at the wrong time of day or while I'm on vacation... I have been digging into how to improve the situation (see my prior comment).

I've some outstanding resources out there that helped me over the hurdles. In case anyone else is interested:
https://docs.microsoft.com/en-us/aspnet/core/security/authentication/identity-primary-key-configuration?tabs=aspnetcore2x
https://www.captechconsulting.com/blogs/Customizing-the-ASPNET-Identity-Data-Model-with-the-Entity-Framework-Fluent-API--Part-1
https://medium.com/@goodealsnow/asp-net-core-identity-3-0-6018fc151b4
and if you're really motivated:
https://www.youtube.com/watch?v=RyActxOsnsg
This last youtube video convinced me that I might as well own the entire setup. Problem solved, and I'm now well-placed to customize away.

**Comment by HaoK** [comment]
This should be now fixed in the latest templates initial migration in 2.1

**Comment by Eilon** [comment]
@HaoK can this bug be closed?

**Comment by AndyMDoyle** [comment]
Out of the box with 2.1.2 we are still seeing the same warnings based on the schema generated with the initial migration.

Removing the initial migration from the template and recreating also results in a schema that throws up these 3 warnings.

Are we missing something?

**Comment by Eilon** [comment]
**Hi, it looks like you are   posting on a closed issue/PR/commit!**

We're very likely to lose track of your bug/feedback/question unless you:

1. Open a new issue
2. Explain very clearly what you need help with
3. If you think you have found a bug, include detailed repro steps so that we can investigate the problem.

**Comment by nanodrive** [comment]
I updated the sdk to 2.1.302 and created a 2.1 web application in visual studio 2017. The number of errors is down to 2 now.

```
Warning! The maximum key length for a clustered index is 900 bytes. The index 'PK_AspNetUserRoles' has maximum length of 1800 bytes. For some combination of large values, the insert/update operation will fail.
Warning! The maximum key length for a clustered index is 900 bytes. The index 'PK_AspNetUserTokens' has maximum length of 1412 bytes. For some combination of large values, the insert/update operation will fail.

```
It's an improvement from 3 errors, but it still needs some more work.

**Comment by coolmikkel** [comment]
> I updated the sdk to 2.1.302 and created a 2.1 web application in visual studio 2017. The number of errors is down to 2 now.
> 
> ```
> Warning! The maximum key length for a clustered index is 900 bytes. The index 'PK_AspNetUserRoles' has maximum length of 1800 bytes. For some combination of large values, the insert/update operation will fail.
> Warning! The maximum key length for a clustered index is 900 bytes. The index 'PK_AspNetUserTokens' has maximum length of 1412 bytes. For some combination of large values, the insert/update operation will fail.
> ```
> It's an improvement from 3 errors, but it still needs some more work.

@nanodrive any news with the last 2 warnings?

I only have the warning regarding the index on AspNetUsersTokens,
Running 2.1.4 sdk.

**Comment by coolmikkel** [comment]
@WellspringCS I do not understand what you are explaining.
Did you change "nvarchar(450)" to "int"?
How do that work? How does a guid gets to be saved in an "int" column?

![image](https://user-images.githubusercontent.com/3486494/46083444-c7c0d900-c1a1-11e8-9f4e-daeb5997644f.png)

**Comment by WellspringCS** [comment]
My bad for not looking back at how I solved the problem last year. On that particular table, I did this:

![image](https://user-images.githubusercontent.com/34347780/46085020-a7295c80-c16a-11e8-9b8c-1c10f3ca94ed.png)

I'm pretty sure in my original solution I had to change UserId from guid to int, and similar changes in several places. But I see you have it as int there, too. Is that the setup out of the box, now?
